README.md
Content Builder
===============

[![Build Status](https://travis-ci.org/Contargo/content-builder.svg?branch=master)](https://travis-ci.org/Contargo/content-builder)
[![](https://jitpack.io/v/Contargo/content-builder.svg)](https://jitpack.io/#Contargo/content-builder)

A small library that provides an easy to use API, supporting creation of
content, in services and back-ends, intended for end-user consumption.

In short - helps you create content in a common format, for web and mobile
clients.

Gettings started
----------------

Using `content-builder` in your project, as a library, means simply including
it as a Maven dependency. We recommend using [Jitpack](https://jitpack.io) to
resolve the dependency:

```xml
<dependency>
  <!-- From Jitpack-repo -->
  <groupId>com.github.Contargo</groupId>
  <artifactId>content-builder</artifactId>
  <version>${SOME-TAG}</version>
</dependency>
```

## General Idea ##

It might sometimes be necessary to create content, for publishing or persisting,
inside a business services or back-end. This means actual readable content, and
not business or domain data. For example creating a message, a note or an event
clarification, that is intended for actual end-users to read.

In order to do this, without coupling each new piece of information created by
the service, to custom code in the client (web/mobile) we've come up with a
small format that we recommend using.

This project provides a small library, that makes it easy to create and consume
such content by services and applications written in Java.

## Micro-Reference ##

The __COLA Common Content Format__ that we use, is easy for developers to
understand and very easy to use. Below is a small example representation in
JSON.

```javascript
{
  contents: [{
      mimeType: 'text/vnd.contargo.description',
      content: 'Kontakt'
    }, {
      mimeType: 'text/vnd.contargo.description',
      locale: 'en',
      content: 'Contact'
    }]
}
```

A collection called `contents` contains one or more content entries. The
required attributes `mimeType` and `content` define the actual information,
with the option to define localized entries using a `locale` attribute.

For more information please see the project Java API documentation and the
available guide- or reference information.

### Vendor Mime/Media-Types ###

Our current choice of content type specification, is to extend on the style and
format of media- or MIME type definitions.

For example `text/vnd.contargo.description` is defined from the pattern:

    '${type}/vnd.contargo.${name}'

With the option to use subtypes and charset definitions, just as with media
types.

    '${type}/vnd.contargo.${name}+${subtype};${params}'

There's more information and concrete definitions in the project Java API
documentation.

## Known Issues

Compatibility with JSON serialization, currently assumes the use of
[Jackson Databind][1]. This might change in the future, to suite more use-
cases.

_We use `jackson-annotations` to ensure omission of empty properties when
serializing to JSON._

 [1]: https://github.com/FasterXML/jackson-databind

## Development ##

This is a pretty straight-forward Java-project, use `mvn` to build,
test and deploy. Happy hacking!


CHANGES.md
Changelog
=========

## v0.4.1

* Minor fixes to the POM, for building on more recent JDKs than 8.

* Add the Spring Boot parent dependency as a base for dependencies and plugins
  used. This aligns the component for better use with projects in and around
  the Spring and Spring Boot eco-system - as is currently intended.

## v0.4.0

* Project setup updated for hosting on Github as an open source project. Adding
  and removing the necessary information and resources.

## v0.3

* Add support for mime types with `params`, this enables content entries with
  variants to be built.
* Changes the semantics for adding empty strings, trimmable empty strings and
  null values. Such content entries are now ignored.

## v0.2

* Add support for TEXT_APPICON and IMAGE_APPICON mime-types.
* Add content accessors for mime-type and locale based value retrieval.
* Add initial support for byte[] content values.

## v0.1

* Initial version with support for text content creation.


src/test/java/net/contargo/content/ContentTest.java
package net.contargo.content;

import com.fasterxml.jackson.databind.ObjectMapper;

import org.junit.Test;

import java.util.Locale;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;


/**
 * @author  Olle Törnström - toernstroem@synyx.de
 */
public class ContentTest {

    @Test
    public void ensureHasPrettyToString() throws Exception {

        Content c1 = new Content("type", "value");
        String string1 = c1.toString();
        assertTrue("Missing information: '" + string1 + "'", string1.contains("mimeType=type"));
        assertTrue("Missing information: '" + string1 + "'", string1.contains("content=value"));
        assertFalse("Must not have information + '" + string1 + "'", string1.contains("locale="));

        Content c2 = new Content("type", "value", Locale.ENGLISH);
        String string2 = c2.toString();
        assertTrue("Missing information: '" + string2 + "'", string2.contains("locale=en"));
    }


    @Test
    public void ensureReadsProperContentFromJSON() throws Exception {

        String json =
            "{\"mimeType\": \"text/vnd.contargo.appicon\", \"content\": \"some-app-icon\", \"locale\": \"sv\"}";

        Content content = new ObjectMapper().readValue(json.getBytes(), Content.class);

        assertEquals(MimeType.TEXT_APPICON_VAL, content.getMimeType());
        assertEquals("some-app-icon", content.getContent().toString());
        assertEquals(new Locale("sv").getLanguage(), content.getLocale().getLanguage());
    }


    @Test
    public void ensureReadsPropertByteContentFromJSON() throws Exception {

        byte[] bytes = new byte[] { 1, 2, 3 };

        String json = Contents.withMimeType(MimeType.IMAGE_APPICON)
                .andValue(bytes)
                .asJSON()
                .replace("[", "")
                .replace("]", "");

        Content content = new ObjectMapper().readValue(json.getBytes(), Content.class);

        assertEquals(MimeType.IMAGE_APPICON_VAL, content.getMimeType());
        assertArrayEquals(bytes, (byte[]) content.getContent());
    }
}


src/test/java/net/contargo/content/ContentsTest.java
package net.contargo.content;

import net.contargo.content.Contents.Buildable;
import net.contargo.content.Contents.Builder;

import org.junit.Test;

import org.skyscreamer.jsonassert.JSONAssert;

import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNotSame;
import static org.junit.Assert.assertTrue;


/**
 * @author  Olle Törnström - toernstroem@synyx.de
 */
public class ContentsTest {

    @Test
    public void ensureContentsBuiltWithSingleSubjectNoLocale() {

        Builder b1 = Contents.withMimeType(MimeType.TEXT_SUBJECT);
        assertNotNull("Missing builder", b1);

        Buildable b2 = b1.andValue("foobar");
        assertNotNull("Missing builder", b2);
        assertNotSame("Should be different", b2, b1);

        List<Content> contents = b2.asList();

        assertNotNull("Missing results", contents);
        assertEquals("Wrong size", 1, contents.size());

        assertContentEquals(contents.get(0), MimeType.TEXT_SUBJECT, "foobar", null);
    }


    private void assertContentEquals(Content content1, MimeType type, String content, Locale locale) {

        assertEquals("Wrong mime type", type.getMimeType(), content1.getMimeType());
        assertEquals("Wrong content", content, content1.getContent());
        assertEquals("Should be empty", locale, content1.getLocale());
    }


    @Test
    public void ensureContentBuiltWithLocalizedDescription() throws Exception {

        List<? extends Content> contents = Contents.withMimeType(MimeType.TEXT_DESCRIPTION)
                .andValue("Hello", Locale.ENGLISH)
                .asList();

        assertNotNull("Missing results", contents);
        assertEquals("Wrong size", 1, contents.size());

        assertContentEquals(contents.get(0), MimeType.TEXT_DESCRIPTION, "Hello", Locale.ENGLISH);
    }


    @Test
    public void ensureTwoValuesForOneMimeType() throws Exception {

        List<Content> contents = Contents.withMimeType(MimeType.TEXT_BODY)
                .andValue("Say it")
                .andValue("Säg det", new Locale("sv"))
                .asList();

        assertNotNull("Missing results", contents);
        assertEquals("Wrong size", 2, contents.size());

        assertContentEquals(contents.get(0), MimeType.TEXT_BODY, "Say it", null);
        assertContentEquals(contents.get(1), MimeType.TEXT_BODY, "Säg det", new Locale("sv"));
    }


    @Test
    public void ensureMultipleMimeTypes() throws Exception {

        List<Content> contents = Contents.withMimeType(MimeType.TEXT_SUBJECT)
                .andValue("Introducing, the magnificent")
                .andWithMimeType(MimeType.TEXT_BODY)
                .andValue("Welcome to the first day of the rest of your life!")
                .andWithMimeType(MimeType.TEXT_DESCRIPTION)
                .andValue("Message")
                .andValue("Nachricht", Locale.GERMAN)
                .asList();

        assertNotNull("Missing results", contents);
        assertEquals("Wrong size", 4, contents.size());

        assertContentEquals(contents.get(0), MimeType.TEXT_SUBJECT, "Introducing, the magnificent", null);
        assertContentEquals(contents.get(1), MimeType.TEXT_BODY, "Welcome to the first day of the rest of your life!",
            null);
        assertContentEquals(contents.get(2), MimeType.TEXT_DESCRIPTION, "Message", null);
        assertContentEquals(contents.get(3), MimeType.TEXT_DESCRIPTION, "Nachricht", Locale.GERMAN);
    }


    @Test
    public void ensureBuildsAsProperMap() throws Exception {

        List<Map<String, Object>> contents = Contents.withMimeType(MimeType.TEXT_BODY)
                .andValue("Say it")
                .andValue("Säg det", new Locale("sv"))
                .asMap();

        assertNotNull("Missing results", contents);
        assertEquals("Wrong size", 2, contents.size());

        assertContentEquals(contents.get(0), MimeType.TEXT_BODY, "Say it", null);
        assertContentEquals(contents.get(1), MimeType.TEXT_BODY, "Säg det", new Locale("sv"));
    }


    private void assertContentEquals(Map<String, Object> entry, MimeType type, String content, Locale locale) {

        assertTrue("Missing required key: mimeType", entry.containsKey("mimeType"));
        assertTrue("Missing required key: content", entry.containsKey("content"));

        assertEquals("Wrong mime type", type.getMimeType(), entry.get("mimeType"));
        assertEquals("Wrong content", content, entry.get("content"));
        assertEquals("Wrong locale", Optional.ofNullable(locale).map(Locale::toString).orElse(null),
            entry.get("locale"));
    }


    @Test
    public void ensureReturnsJsonString() throws Exception {

        String json = Contents.withMimeType(MimeType.TEXT_BODY)
                .andValue("Say it")
                .andValue("Säg det", new Locale("sv"))
                .asJSON();

        JSONAssert.assertEquals("["
            + "{mimeType: 'text/vnd.contargo.body', content: 'Say it'}, "
            + "{mimeType: 'text/vnd.contargo.body', content: 'Säg det', locale: 'sv'}]", json, true);
    }


    @Test
    public void ensureRetrievesContentByMimeType() throws Exception {

        List<Content> contents = Contents.withMimeType(MimeType.TEXT_BODY)
                .andValue("Say it")
                .andValue("Säg det", new Locale("sv"))
                .asList();

        String value = new Contents(contents).forMimeType(MimeType.TEXT_BODY);

        assertEquals("Say it", value);
    }


    @Test
    public void ensureMatchesVariants() throws Exception {

        List<Content> contents = Contents.withMimeType(MimeType.TEXT_BODY.withParams("foo"))
                .andValue("variant")
                .andWithMimeType(MimeType.TEXT_BODY)
                .andValue("default")
                .asList();

        String v1 = new Contents(contents).forMimeType(MimeType.TEXT_BODY);
        String v2 = new Contents(contents).forMimeType(MimeType.TEXT_BODY.withParams("foo"));

        assertEquals("default", v1);
        assertEquals("variant", v2);
    }


    @Test
    public void ensureRetreivesContentByMimeTypeAndLocale() throws Exception {

        List<Content> contents = Contents.withMimeType(MimeType.TEXT_BODY)
                .andValue("Say it")
                .andValue("Säg det", new Locale("sv"))
                .asList();

        String value = new Contents(contents).forMimeTypeAndLocale(MimeType.TEXT_BODY, new Locale("sv"));

        assertEquals("Säg det", value);
    }


    @Test
    public void ensureCanCreateAndRetriveBinaryImageData() throws Exception {

        byte[] image = new byte[] { 1, 2, 3 };
        List<Content> contents = Contents.withMimeType(MimeType.IMAGE_APPICON).andValue(image).asList();

        byte[] value = new Contents(contents).forMimeType(MimeType.IMAGE_APPICON);
        assertArrayEquals(image, value);
    }


    @Test
    public void ensureIgnoresEmptyContentEntries() throws Exception {

        List<Content> contents = Contents.withMimeType(MimeType.TEXT_APPICON)
                .andValue("    ")
                .andWithMimeType(MimeType.TEXT_SUBJECT)
                .andValue("")
                .andWithMimeType(MimeType.TEXT_BODY)
                .andValue((String) null)
                .andWithMimeType(MimeType.TEXT_DESCRIPTION)
                .andValue("foo")
                .asList();

        assertNotNull("Missing results", contents);
        assertEquals("Wrong size", 1, contents.size());
        assertContentEquals(contents.get(0), MimeType.TEXT_DESCRIPTION, "foo", null);
    }
}


src/test/java/net/contargo/content/MimeTypeTest.java
package net.contargo.content;

import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;


/**
 * @author  Olle Törnström - toernstroem@synyx.de
 */
public class MimeTypeTest {

    @Test
    public void ensureDeclaredContentMimeTypeConstants() {

        assertEquals("text/vnd.contargo.subject", MimeType.TEXT_SUBJECT.getMimeType());
        assertEquals("text/vnd.contargo.description", MimeType.TEXT_DESCRIPTION.getMimeType());
        assertEquals("text/vnd.contargo.body", MimeType.TEXT_BODY.getMimeType());
    }


    @Test
    public void ensureEqualsForInstancesAreCorrect() throws Exception {

        assertTrue(MimeType.TEXT_SUBJECT.equals(MimeType.TEXT_SUBJECT));
        assertTrue(MimeType.TEXT_DESCRIPTION.equals(MimeType.TEXT_DESCRIPTION));

        assertTrue(MimeType.TEXT_BODY.equals(MimeType.TEXT_BODY));
        assertFalse(MimeType.TEXT_BODY.equals(MimeType.TEXT_DESCRIPTION));
        assertFalse(MimeType.TEXT_BODY.equals(MimeType.TEXT_SUBJECT));
    }


    @Test
    @SuppressWarnings("unlikely-arg-type")
    public void ensureEqualsForStringComparisonIsAvailable() throws Exception {

        assertTrue(MimeType.TEXT_SUBJECT.equals("text/vnd.contargo.subject"));
        assertFalse(MimeType.TEXT_SUBJECT.equals("text/vnd.contargo.description"));

        assertTrue(MimeType.TEXT_DESCRIPTION.equals("text/vnd.contargo.description"));
        assertTrue(MimeType.TEXT_BODY.equals("text/vnd.contargo.body"));
    }


    @Test
    public void ensureCanCreateMimeTypeVariantsWithParams() throws Exception {

        assertEquals("text/vnd.contargo.appicon;foobar", MimeType.TEXT_APPICON.withParams("foobar").getMimeType());
    }
}


src/main/java/net/contargo/content/Content.java
package net.contargo.content;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;

import java.text.MessageFormat;

import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;


/**
 * A COLA Common Content Format entry.
 *
 * <p>This class represents the unit of a content entry, in the COLA Common Content Format. It will contain the
 * mime-type description and content data, with an optional language or locale, for localization.</p>
 *
 * @author  Olle Törnström - toernstroem@synyx.de
 * @since  0.1
 */
@JsonInclude(Include.NON_ABSENT)
@JsonDeserialize(using = ContentDeserializer.class)
public class Content {

    private String mimeType;
    private Object content;
    private Optional<Locale> locale;

    protected Content() {

        // Required!
    }


    Content(String mimeType, Object content) {

        this(mimeType, content, Optional.empty());
    }


    Content(String mimeType, Object content, Locale locale) {

        this(mimeType, content, Optional.of(locale));
    }


    /**
     * Private constructor for content entries.
     *
     * @param  mimeType  of the entry
     * @param  content  of the entry
     * @param  locale  of the entry
     */
    private Content(String mimeType, Object content, Optional<Locale> locale) {

        this.mimeType = mimeType;
        this.content = content;
        this.locale = locale;
    }

    /**
     * Returns the mime type of this content entry.
     *
     * @return  mime type string
     */
    public String getMimeType() {

        return this.mimeType;
    }


    /**
     * Returns the content data of this entry.
     *
     * @return  content data object
     */
    public Object getContent() {

        return content;
    }


    /**
     * Returns the locale of this content entry.
     *
     * @return  the entry locale or {@code null} if it does not exist
     */
    public Locale getLocale() {

        return locale.orElse(null);
    }


    @Override
    public String toString() {

        if (locale.isPresent()) {
            return MessageFormat.format("Content [mimeType={0}, content={1}, locale={2}]", mimeType, content,
                    locale.get());
        }

        return MessageFormat.format("Content [mimeType={0}, content={1}]", mimeType, content);
    }


    static Map<String, Object> asMap(Content content) {

        Map<String, Object> map = new HashMap<>(3);

        map.put("mimeType", content.getMimeType());
        map.put("content", content.getContent());

        Optional.ofNullable(content.getLocale()).ifPresent(l -> map.put("locale", l.toString()));

        return map;
    }


    boolean forMimeType(MimeType mimeType) {

        return this.mimeType.equals(mimeType.getMimeType());
    }


    boolean forMimeTypeAndLocale(MimeType mimeType, Locale locale) {

        String lang = this.locale.map(Locale::getLanguage).orElse(null);

        return locale.getLanguage().equals(lang) && this.mimeType.equals(mimeType.getMimeType());
    }
}


src/main/java/net/contargo/content/Contents.java
package net.contargo.content;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;


/**
 * Chaining builder for COLA Common Content, providing an easy to use and developer-friendly API for working with
 * content creation.
 *
 * @author  Olle Törnström - toernstroem@synyx.de
 * @since  0.1
 */
public final class Contents {

    private final List<Content> values = new ArrayList<>();

    private Contents() {

        // Hidden
    }


    /**
     * Creates a new contents instance with the given values.
     *
     * @param  values  to append to the new contents instance
     */
    public Contents(List<Content> values) {

        this.values.addAll(Optional.ofNullable(values).orElse(Collections.emptyList()));
    }

    /**
     * Creates a new contents builder, starting it of with the given mime-type.
     *
     * @param  mimeType  to start building with
     *
     * @return  a new contents builder instance
     */
    public static Builder withMimeType(MimeType mimeType) {

        return new ContentBuilder(mimeType, new Contents());
    }


    /**
     * Retrieves the content matching the given MIME-type.
     *
     * @param  mimeType  predicate to match
     * @param  <T>  inferring the type of content to retrieve
     *
     * @return  the content value found, or {@code null} if not content could be found
     *
     * @since  0.2
     */
    @SuppressWarnings("unchecked")
    public <T> T forMimeType(MimeType mimeType) {

        return (T) this.values.stream().filter(c -> c.forMimeType(mimeType)).findFirst().map(Content::getContent)
            .orElse(null);
    }


    /**
     * Retrieves the content matching the given MIME-type and locale.
     *
     * @param  mimeType  predicate to match
     * @param  locale  predicate to match
     * @param  <T>  inferring the type of content to retrieve
     *
     * @return  the content value found, or {@code null} if not content could be found
     *
     * @since  0.2
     */
    @SuppressWarnings("unchecked")
    public <T> T forMimeTypeAndLocale(MimeType mimeType, Locale locale) {

        return (T) this.values.stream().filter(c -> c.forMimeTypeAndLocale(mimeType, locale)).findFirst()
            .map(Content::getContent)
            .orElse(null);
    }

    /**
     * Builder can add value, with an optional locale.
     *
     * @author  Olle Törnström - toernstroem@synyx.de
     * @since  0.1
     */
    public interface Builder {

        /**
         * Adds a value to the builder.
         *
         * @param  value  to add
         *
         * @return  a buildable builder
         */
        Buildable andValue(String value);


        /**
         * Adds a value to the builder with a given locale.
         *
         * @param  value  to add
         * @param  locale  of the value localization/language
         *
         * @return  a buildable builder
         */
        Buildable andValue(String value, Locale locale);


        /**
         * Add a value to the builder for the current default locale.
         *
         * @param  value  to add
         *
         * @return  a buildable builder
         */
        Buildable andValue(byte[] value);
    }

    /**
     * Appendable can append another mime-type content entry to the builder.
     *
     * @author  Olle Törnström - toernstroem@synyx.de
     * @since  0.1
     */
    public interface Appendable extends Builder {

        /**
         * Appends the given mime type to the builder.
         *
         * @param  mimeType  to append
         *
         * @return  a builder
         */
        Builder andWithMimeType(MimeType mimeType);
    }

    /**
     * Buildable can build the resulting contents.
     *
     * @author  Olle Törnström - toernstroem@synyx.de
     * @since  0.1
     */
    public interface Buildable extends Appendable {

        /**
         * Builds the resulting contents as a list of strongly typed content entries.
         *
         * @return  a list of content entries
         */
        List<Content> asList();


        /**
         * Builds the resulting content as a collection of maps, reducing avoiding the content type to be bound to the
         * instance.
         *
         * @return  a list of map entries
         */
        List<Map<String, Object>> asMap();


        /**
         * Builds the resulting content as a JSON string.
         *
         * @return  a JSON string
         */
        String asJSON();
    }

    private static final class ContentBuilder implements Buildable {

        private final Contents contents;
        private final MimeType mimeType;

        private ContentBuilder(MimeType mimeType, Contents contents) {

            this.mimeType = mimeType;
            this.contents = contents;
        }


        private ContentBuilder(ContentBuilder contentBuilder, String value) {

            this.mimeType = contentBuilder.mimeType;
            this.contents = contentBuilder.contents;

            Content content = new Content(this.mimeType.getMimeType(), value);
            this.contents.values.add(content);
        }


        public ContentBuilder(ContentBuilder contentBuilder, byte[] value) {

            this.mimeType = contentBuilder.mimeType;
            this.contents = contentBuilder.contents;

            Content content = new Content(this.mimeType.getMimeType(), value);
            this.contents.values.add(content);
        }


        private ContentBuilder(ContentBuilder contentBuilder, String value, Locale locale) {

            this.mimeType = contentBuilder.mimeType;
            this.contents = contentBuilder.contents;

            Content content = new Content(this.mimeType.getMimeType(), value, locale);
            this.contents.values.add(content);
        }

        @Override
        public Buildable andValue(String value) {

            return isNullOrEmpty(value) ? this : new ContentBuilder(this, value);
        }


        private boolean isNullOrEmpty(String value) {

            return value == null || value.trim().isEmpty();
        }


        @Override
        public Buildable andValue(byte[] value) {

            return isNullOrSizeZero(value) ? this : new ContentBuilder(this, value);
        }


        private boolean isNullOrSizeZero(byte[] value) {

            return value == null || value.length == 0;
        }


        @Override
        public Buildable andValue(String value, Locale locale) {

            return isNullOrEmpty(value) ? this : new ContentBuilder(this, value, locale);
        }


        @Override
        public Builder andWithMimeType(MimeType mimeType) {

            return new ContentBuilder(mimeType, contents);
        }


        @Override
        public List<Content> asList() {

            return Collections.unmodifiableList(new ArrayList<>(this.contents.values));
        }


        @Override
        public List<Map<String, Object>> asMap() {

            return this.contents.values.stream().map(Content::asMap).collect(Collectors.toList());
        }


        @Override
        public String asJSON() {

            try {
                return new ObjectMapper().writeValueAsString(asMap());
            } catch (JsonProcessingException e) {
                throw new RuntimeException("Could not write contents as JSON string", e);
            }
        }
    }
}


src/main/java/net/contargo/content/ContentMimeType.java
package net.contargo.content;

/**
 * Provides the declared constant content mime-types.
 *
 * <p>Our use of a {@code mimeType} property, is how we like to classify and structure content. This class is an attempt
 * to define the currently known and supported declarations.</p>
 *
 * <p>Though not strictly IANA or RFC compatible, we have chosen to keep to a structure that expresses <i>some
 * compatibility</i> (and future safety) with said standards. Please note that it's currently also <strong>not</strong>
 * the intent to publish these type definitions, outside of the COLA system landscape.</p>
 *
 * <p>Our mime type definitions are defined using the following pattern, with optional variations:</p>
 *
 * <pre><code>
${type}/vnd.contargo.${name}
${type}/vnd.contargo.${name};charset=${charset}
${type}/vnd.contargo.${name}+${subtype}
${type}/vnd.contargo.${name}+${subtype};charset=${charset}
   </code></pre>
 *
 * <p>The constant {@code vnd.contargo} is our vendor specifier.</p>
 *
 * @author  Olle Törnström - toernstroem@synyx.de
 * @since  0.1
 */
public final class ContentMimeType {

    private static final String _P = "text/vnd.contargo.";

    /**
     * A subject should describe and give context to some content in a single line of text.
     */
    public static final String TEXT_SUBJECT = _P + "subject";

    /**
     * The description should, in a summary or short amount of text, give information about some content.
     */
    public static final String TEXT_DESCRIPTION = _P + "description";

    /**
     * Some text that represents the full content of some information.
     */
    public static final String TEXT_BODY = _P + "body";
}


src/main/java/net/contargo/content/MimeType.java
package net.contargo.content;

import java.util.Objects;


/**
 * The declared constant content mime-types.
 *
 * <p>Our use of a {@code mimeType} property, is how we like to classify and structure content. This class is an
 * attempt to define the currently known and supported declarations. By that I mean, since it's basically <i>
 * possible</i> to use whatever string value you want, this is merely a list of the <strong>recommended</strong>
 * ones.</p>
 *
 * <p>Though not strictly IANA or RFC compatible, we have chosen to keep to a structure that expresses <i>some
 * compatibility, and future safety,</i> with said standards. Please note that it's currently also <strong>not</strong>
 * the intent to publish the type definitions here, as official public extensions, outside of it's use in this library.
 * </p>
 *
 * <p>Our mime type definitions are defined using the following pattern, with optional variations:</p>
 *
 * <pre><code>
${type}/vnd.contargo.${name}
${type}/vnd.contargo.${name};${params}
${type}/vnd.contargo.${name}+${subtype}
${type}/vnd.contargo.${name}+${subtype};${params}
   </code></pre>
 *
 * <p>The constant {@code vnd.contargo} is our vendor specifier. Which for our purposes is <strong>mandatory</strong>.
 * The {@code type}, {@code subtype} and {@code params} values, try to follow the way that IANA/RFC based media types
 * are declared.</p>
 *
 * <p>What we describe with {@code name} should hopefully have some semantic but still be open for broad use. We always
 * try to define these types with documentation, as a hint for developers.</p>
 *
 * <p>Here's an example of what a mime type could look like: <code>text/vnd.contargo.appicon+uri;class=small</code></p>
 *
 * <p>In this example a mime-type <em>parameter</em> is used to mark the content with a <code>class</code>, this is not
 * a specific parameter but an option for the authoring code, as means to extend the specification of the provided
 * content, to the user.</p>
 *
 * @author  Olle Törnström - toernstroem@synyx.de
 * @since  0.1
 */
public final class MimeType {

    /**
     * @see  #TEXT_SUBJECT
     */
    public static final String TEXT_SUBJECT_VAL = "text/vnd.contargo.subject";

    /**
     * @see  #TEXT_DESCRIPTION
     */
    public static final String TEXT_DESCRIPTION_VAL = "text/vnd.contargo.description";

    /**
     * @see  #TEXT_BODY
     */
    public static final String TEXT_BODY_VAL = "text/vnd.contargo.body";

    /**
     * @see  #TEXT_APPICON
     */
    public static final String TEXT_APPICON_VAL = "text/vnd.contargo.appicon";

    /**
     * @see  #IMAGE_APPICON
     */
    public static final String IMAGE_APPICON_VAL = "image/vnd.contargo.appicon";

    /**
     * A subject should describe and give context to some content in a single line of text: <code>
     * {@value #TEXT_SUBJECT_VAL}</code>.
     *
     * <p>For example an email or message subject line, describing the contents.</p>
     */
    public static final MimeType TEXT_SUBJECT = MimeType.of(TEXT_SUBJECT_VAL);

    /**
     * The description should, in a summary or short amount of text, give information about some content: <code>
     * {@value #TEXT_DESCRIPTION_VAL}</code>.
     *
     * <p>For example a clarification on some piece of information, such as it's type, structure or classification -
     * currently used to describe the domain type of a search-result.</p>
     */
    public static final MimeType TEXT_DESCRIPTION = MimeType.of(TEXT_DESCRIPTION_VAL);

    /**
     * Some text that represents the full content of some information: <code> {@value #TEXT_BODY_VAL}</code>
     *
     * <p>For example a message body, from a service or machine, that has to be sent and persisted in multiple
     * languages.</p>
     */
    public static final MimeType TEXT_BODY = MimeType.of(TEXT_BODY_VAL);

    /**
     * Describes an application icon image resource reference, such as a path or a URI: <code>
     * {@value #TEXT_APPICON_VAL}</code>.
     *
     * <p>Typically provided for the rendering of a size-independent (scalable) image or background. More specifically,
     * the URI to an application's SVG icon resource.</p>
     *
     * <p>NOTE: It is intended for the content author to optionally extend the content with a classification, which may
     * allow for more suitable use by the client. Such classifications are preferably to apply with a mime-type
     * parameter, for example: {@code ;class=small}</p>
     *
     * @since  0.2
     */
    public static final MimeType TEXT_APPICON = MimeType.of(TEXT_APPICON_VAL);

    /**
     * Describes an application binary image icon <code>{@value #IMAGE_APPICON_VAL}</code>.
     *
     * @since  0.2
     */
    public static final MimeType IMAGE_APPICON = MimeType.of(IMAGE_APPICON_VAL);

    private final String mimeType;

    /**
     * Constructs a new mime type with the given value.
     *
     * <p>Absolutely <strong>no checks or validations made</strong>. The provided mime type string will be used
     * as-is.</p>
     *
     * @param  mimeType  value to create
     */
    public MimeType(String mimeType) {

        this.mimeType = mimeType;
    }

    /**
     * Returns the mime type value for this instance.
     *
     * @return  mime type value as string
     */
    String getMimeType() {

        return mimeType;
    }


    /**
     * Private static factory class, used for constant static mime type declarations.
     *
     * @param  mimeTypeValue  string for the mime type to create
     *
     * @return  a new instance of a mime type with the given value
     */
    private static MimeType of(String mimeTypeValue) {

        return new MimeType(mimeTypeValue);
    }


    @Override
    public int hashCode() {

        return Objects.hashCode(mimeType);
    }


    @Override
    public boolean equals(Object obj) {

        if (this == obj) {
            return true;
        }

        if (obj == null) {
            return false;
        }

        if (obj instanceof String) {
            return Objects.equals(mimeType, obj);
        }

        if (obj instanceof MimeType) {
            return Objects.equals(mimeType, ((MimeType) obj).mimeType);
        }

        return false;
    }


    /**
     * Appends the given parameters to this mime-type.
     *
     * @param  params  to append
     *
     * @return  the mime type with the given parameters appended after the semicolon ({@code ;}) separator character
     */
    public MimeType withParams(String params) {

        return new MimeType(String.format("%s;%s", mimeType, params));
    }
}
